### ISSUE TITLE
Project published on Windows fails to load System.Net.Security on Linux

### ISSUE THREAD

**Issue by fsoikin** [aspnet/dnx#3272]
**TL;DR**: if my program uses `System.Net.Security`, and I `dnu publish --no-source` it on a Windows machine, and then try to run that published version on a Linux machine, it fails to load `System.Net.Security`.
Plus, it hangs up trying to load it, for which I have filed a separate issue [#2429 in coreclr](https://github.com/dotnet/coreclr/issues/2429), because the hangup is not specific to that particular library.

Below is a listing of my actions, and here are some details:
- The Windows machine is Windows 10 x64.
- Linux machine is CentOS 7.2 x64.
- Runtime is `rc1-update1 -r coreclr` (on both machines), but this also reproduces with `rc2-16317`.
- Running `dnu restore` on Linux before executing the program lets `System.Net.Security` load, but breaks in another way, see #3273.

```
C:\i\c> ls

    Directory: C:\i\c

Mode                LastWriteTime         Length Name
----                -------------         ------ ----
-a----       12/19/2015  11:52 AM            158 Program.cs
-a----       12/19/2015  11:52 AM            295 project.json

C:\i\c> cat .\project.json
{
        "compilationOptions": {
                "emitEntryPoint": true
        },

        "commands": {
                "c": "c"
        },

        "dependencies": {
                "System.Runtime": "4.0.20-beta-23409",
                "System.Console": "4.0.0-beta-23409",
                "System.Net.Security": "4.0.0-beta-23409"
        },

        "frameworks": {
                "dnxcore50": { }
        }
}

C:\i\c> cat .\Program.cs
public static class Program {
        public static void Main( string[] args ) {
                System.Console.WriteLine( typeof( System.Net.Security.SslStream ).Name );
        }
}

C:\i\c> dnu restore
<snip>

C:\i\c> dnx run
SslStream

C:\i\c> dnu publish --no-source --out C:\work\CentOS\c --quiet
Microsoft .NET Development Utility CoreClr-x64-1.0.0-rc1-16231

Copying to output path C:\work\CentOS\c
Time elapsed 00:00:01.6781552

C:\i\c>

*************************************
***** switch to CentOS 7.2 bash *****
*************************************

[fsoikin@CentOS CentOS]$ c/approot/c
<hang>
^C

[fsoikin@CentOS CentOS]$ DNX_TRACE=1 c/approot/c
Loaded module: dnx.coreclr.so
Found export: CallApplicationMain
Information: [DomainManager] Using CoreCLR
Information: [LoaderContainer]: Load name=Microsoft.Dnx.ApplicationHost, Culture=neutral, PublicKeyToken=null
Information: [PathBasedAssemblyLoader]: Loaded name=Microsoft.Dnx.ApplicationHost, Culture=neutral, PublicKeyToken=null in 21ms
Information: [Bootstrapper] Runtime Framework: DNXCore,Version=v5.0
Information: [LoaderContainer]: Load name=Microsoft.Dnx.Runtime, Version=1.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60
Information: [PathBasedAssemblyLoader]: Loaded name=Microsoft.Dnx.Runtime, Version=1.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60 in 0ms
Information: [LoaderContainer]: Load name=Microsoft.Dnx.Compilation, Version=1.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60
Information: [PathBasedAssemblyLoader]: Loaded name=Microsoft.Dnx.Compilation, Version=1.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60 in 0ms
Information: [LoaderContainer]: Load name=Microsoft.Dnx.Compilation.Abstractions, Version=1.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60
Information: [PathBasedAssemblyLoader]: Loaded name=Microsoft.Dnx.Compilation.Abstractions, Version=1.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60 in 0ms
Information: [LoaderContainer]: Load name=System.Threading.Thread, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
Information: [PathBasedAssemblyLoader]: Loaded name=System.Threading.Thread, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a in 0ms
Warning: [ApplicationHostContext] Lock File Target is Runtime-agnostic! This is generally not good.
Information: [ApplicationHostContext] Using Lock File Target: DNXCore,Version=v5.0/
Information: [DefaultHost]: Project path: /mnt/hgfs/CentOS/c/approot/packages/c/1.0.0/root
Information: [DefaultHost]: Project root: /mnt/hgfs/CentOS/c/approot
Information: [DefaultHost]: Project configuration: Debug
Information: [DefaultHost]: Packages path: /mnt/hgfs/CentOS/c/approot/packages
Information: [Breadcrumbs] Breadcrumbs for servicing will not be written because the breadcrumbs folder () does not exist.
Information: [LoaderContainer]: Load name=System.Threading.ThreadPool, Version=4.0.10.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
Information: [PathBasedAssemblyLoader]: Loaded name=System.Threading.ThreadPool, Version=4.0.10.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a in 0ms
Information: [LoaderContainer]: Load name=c, Culture=neutral, PublicKeyToken=null
Information: [PackageAssemblyLoader]: Loaded name=c, Culture=neutral, PublicKeyToken=null in 4ms
Information: [LoaderContainer]: Load name=System.Net.Security, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
<hang>
^C

[fsoikin@CentOS CentOS]$ 
```

**Comment by BrennanConroy** [comment]
You have to restore on the windows machine passing in the linux runtime id so that it restores the linux specific dependencies

**Comment by BrennanConroy** [comment]
```dnu restore --runtime centos.7.1-x64```

**Comment by fsoikin** [comment]
Do you mean to do that _before_ I publish?

**Comment by BrennanConroy** [comment]
Actually, you should install the dnx you want to publish with and do ```dnu publish --runtime <linux dnx>``` and that should do everything I believe

**Comment by fsoikin** [comment]
I don't quite understand how to identify "linux dnx" vs "windows dnx", as they have same moniker. Can you give me an example of full command?

**Comment by BrennanConroy** [comment]
```dnu publish --out <path> --runtime dnx-coreclr-linux-x64```
vs.
```dnu publish --out <path> --runtime dnx-coreclr-win-x64```

unfortunately you're trying centos which will create a lot of difficulties for you since we didn't make a centos dnx

**Comment by fsoikin** [comment]
I have the linux one installed on CentOS, so that should be ok, I'll try it out.

But now, this raises next question: since I have to know target platform in advance, this is not "cross-platform" anymore, is it?

Is there a way to package my app so it would run anywhere?

**Comment by fsoikin** [comment]
Ok, tried that out.

```
C:\i\c> dnu publish --no-source --out C:\work\CentOS\c --runtime dnx-coreclr-linux-x64
Microsoft .NET Development Utility CoreClr-x64-1.0.0-rc1-16231

The project being published does not support the runtime 'dnx-coreclr-linux-x64'
```

**Comment by fsoikin** [comment]
Ok, after reading docs for `dnu publish`, I decided to try with full path to the runtime, and publish went through fine.

```
dnu publish --no-source --out C:\work\CentOS\c --runtime C:\Users\Fyodor\.dnx\runtimes\dnx-coreclr-linux-x64.1.0.0-rc1-update1
```

But then I switched to CentOS and tried to run it, and guess what? Same result!

```
[Fyodor@CentOS CentOS] $ c/approot/c
<hang>

# But if I run "dnu restore" on Linux now, it starts working
[Fyodor@CentOS CentOS] $ cd c/approot
[Fyodor@CentOS CentOS] $ dnu restore
<snip>

[Fyodor@CentOS CentOS] $ ./c
SslStream

[Fyodor@CentOS CentOS] $ 
```
